Groundcover consolidation

bosvensson1 opened the pull request:
Thanks to the suggestions and help from psi 29a I have made groundcover more efficient at a high level. With these changes groundcover reuses the existing QuadTreeWorld. This reduces the calculations required for groundcover and allows preloading. The setting for a larger min chunk size for groundcover was consequently removed in the process. The larger chunk size I assume was only used because larger chunks do not have to reload as often as smaller ones as we move around. With loading times a non-issue, we are better off with smaller chunks allowing more of the off screen groundcover to be culled.

AbduSharif commented:
You might also want to rove this from the documentation once it's reviewed:
https://openmw.readthedocs.io/en/latest/reference/modding/settings/groundcover.html#min-chunk-size

psi29a commented:
Pulling in @akortunov and @AnyOldName3 to help review.

It's very cool of you to take this up. :)

We've been discussing further here: 
https://gitlab.com/OpenMW/openmw/-/issues/6263

akortunov commented:
From what I can tell, with new approach chunk get invalidated every time when ANY of its groundcover instances enters or leaves viewing distance. For me it means 10-20 chunk rebuilds every second when running with normal speed.

Are you sure that it is a good idea?

akortunov commented:
> I assume was only used because larger chunks do not have to reload as often as smaller ones as we move around

No, the setting was added because these is a balance between CPU vs GPU usage which can be different of different setups. With larger chunks you render some offscreen vertices (more job for GPU), but you get a reduced amount of draw calls and faster culling.

AnyOldName3 commented:
The vertex shader is also pretty cheap, so having out-of-frustum grass objects doesn't cause problems as quickly as having lots of draw calls. It's possible that was partially affected by some other issue this resolves (e.g. if this reduces state switches between draw calls compared to before) as I've not looked at the diff yet, so the effect might be mitigated, but I'd be surprised. Even simply getting the vertex shader to move vertices belonging to grass instances beyond the grass distance way outside the frustum had a significant impact, so it's not reasonable to equate an extra hundred grass instances outside the frustum with the hundred already visible when estimating performance.

Basically, before decreasing the grass chunk size, we should actually benchmark to see if the extra culling is worth any extra state switches and draw calls as I'm pretty sure the existing size was picked based on data rather than guessing.

AnyOldName3 commented:
Yeah, this looks reasonable except potentially the chunk size setting removal and rebuild-happy chunks. Hopefully, we can gather data to prove the new fixed size is optimal for everyone (including on a range of hardware with a range of grass densities with more than one grass mod) or partially decouple how deep the quad tree gets for terrain and groundcover so we can bring the setting back.

AbduSharif commented:
The game crashed for me at default settings with groundcover enabled:
[Bug report.txt](https://github.com/OpenMW/openmw/files/7124860/Bug.report.txt)

I'm using OSGoS 3.6 (openmw's fork of OSG).

AbduSharif commented:
And yeah it crashed at engine initialization, Here's where it crashed exactly:
[openmw.log](https://github.com/OpenMW/openmw/files/7124872/openmw.log)

akortunov commented:
> From what I can tell, with new approach groundcover chunk get invalidated every time when ANY of its groundcover instances enters or leaves viewing distance. For me it means 10-20 chunk rebuilds every second when running with normal speed.

Object Paging is also affected by this issue with this PR.

akortunov commented:
> And yeah it crashed at engine initialization, Here's where it crashed exactly

Is there a possibility that the crash is caused by 3d-party patches (such as terrain occlusion)?

AbduSharif commented:
> > And yeah it crashed at engine initialization, Here's where it crashed exactly
> 
> Is there a possibility that the crash is caused by 3d-party patches (such as terrain occlusion)?

I'm not using any 3rd party patches other than openmw-android's this one which allows precompiled shader archive gl4es feature work with opemw:
https://github.com/Sisah2/openmw-android/blob/my/buildscripts/patches/openmw/psa.patch
I also disabled the feature and the crash still happened.

Other than that it's just this PR pulled.

akortunov commented:
An example of chunk size comparison (GTX1050 + i7700):

https://imgsli.com/NzAwNDk

akortunov commented on components/terrain/viewdata.cpp:
```diff
@@ -162,11 +162,6 @@ ViewData *ViewDataMap::getViewData(osg::Object *viewer, const osg::Vec3f& viewPo
             vd->copyFrom(*mostSuitableView);
             return vd;
         }
-        else if (!mostSuitableView)
```
With this change QuadTreeWorld becomes allowed to reuse chunks from different distances instead of building approproate ones. This causes grass pop-ins, when QuadTreeWorld decides to use an empty chunk (from distance beyound rendering distance) for nearby objects instead of building a chunk with data.

bosvensson1 replied:
The issue you experienced is the logical consequence when there is no preloading in place as the code copying across views assumed. If we do not experience this anymore the fix is not necessary if not counterproductive.

akortunov replied:
Well, with this PR this issue exists and pop-ints actually happen.

bosvensson1 replied:
Then it seems we have other issues the mentioned change hides.

In theory, you should only see pop-in when the preloader is too slow. In this case, would you have rather have pop-in or a loading break?

In practice, the preloader uses a predicted position 1 second in the future. It is usually not the real position at the time of loading. In my experience, the preloader is much faster and the 1 second value is too high. The time should be adjusted by the actual loading time or there should be a distance threshold to mitigate the inaccuracy. Maybe both. I will try to add a threshold, see if this is enough. The impact of this additional loading distance is minor as Any old name explains, the vertices discarded by the shader have little impact.

akortunov replied:
> In theory, you should only see pop-in when the preloader is too slow. In this case, would you have rather have pop-in or a loading break?

If I understood correctly, it does not work like that. QuadTreeWorld has zero reasons to render new chunk  immediately as it becomes loaded. It will be rendered only when it will be requested in the next time (e.g. during cell change).

bosvensson1 replied:
No there is a reason. We call getViewData every frame https://github.com/OpenMW/openmw/blob/44d1a77a875a6211ae51809cb43fd0d251a5cba7/components/terrain/quadtreeworld.cpp#L436 which copies in the most appropriate view available (usually one made by the preloader) if the current view is beyond reusedistance.
https://github.com/bosvensson1/openmw/blob/726c2ac7adbcc64dfed507f729205b48a05efcae/components/terrain/viewdata.cpp#L144

At the moment though the check for rendering distance does not account for reusedistance. Will try to add that.

bosvensson1 commented:
> And yeah it crashed at engine initialization, Here's where it crashed exactly:
> [openmw.log](https://github.com/OpenMW/openmw/files/7124872/openmw.log)

Are you using groundcover without distant terrain on? The new approach depends on distant being on. I forgot to enforce this in the code. The old approach did not mandate distant terrain to be on but did most of the setup work involved with distant terrain regardless. A slight misnomer setting wise.

MocquillonCedric commented:
Hi. For the min chunk size, I don't know if it could help or not but I made some experiments in the [past](https://discord.com/channels/260439894298460160/260443767482482688/862259356342288394) with changing the default value for the quadtreeworld to [0.5](https://gitlab.com/CedricMocquillon/openmw/-/commit/8bd03205441d802bafffb8788103bfafeb9963c1) (i.e. the same as the default one for akortunov' grass groundcover implementation) and I and CMAugust found that it increases the fps (https://discord.com/channels/260439894298460160/260443767482482688/862260772574855208)

So perhaps something greater than 1/8 could be the "new fixed size is optimal for everyone" (sure we will have to made other experiments to prove that)

akortunov commented:
> A slight misnomer setting wise.

It is quite strange that nearby groundcover starts to depend on *distant* terrain.

AbduSharif commented:
> The new approach depends on distant being on. 

What would be strange is needing to enable two settings to enable groundcover.

bosvensson1 commented:
> Yeah, this looks reasonable except potentially the chunk size setting removal and rebuild-happy chunks. Hopefully, we can gather data to prove the new fixed size is optimal for everyone (including on a range of hardware with a range of grass densities with more than one grass mod) or partially decouple how deep the quad tree gets for terrain and groundcover so we can bring the setting back.

It is too early for those benchmarks. I only get to the highest level of this performance work in this PR. I will work on another PR addressing low level bottlenecks that will change the situation in benchmarks more drastically than the higher level work. The higher level work isn't supposed to improve FPS performance, it is supposed to improve smoothness with preloading and reduced setup costs as explained in the first post.

The numbers shown by akortunov show that while the FPS are the same, the smaller chunks presently necessitate more CPU work in the frame. This surprised me. There should be nothing inherent to groundcover tipping the scales in the direction of a CPU bottleneck compared to terrain and objects. The opposite should rather be the case. It seems we currently have issues because each groundcover chunk uses osg::clone with aggressive parameters and creates its own unique state. osg can not see that the different chunks use almost the same state. We can easily improve on this with a shallower copy and caches where necessary. This is what I plan to do in a next PR after this one. If there is still a benefit to larger chunks after this next PR, likely the terrain can benefit from larger chunks as well. Then we can expose the chunk size generally. But I assume there must be some reason why the terrain chunk size is not configurable. Does anyone know why?

akortunov commented:
> The opposite should rather be the case. It seems we currently have issues because each groundcover chunk uses osg::clone with aggressive parameters and creates its own unique state. osg can not see that the different chunks use almost the same state.

The problem is that instances count is the part of shape buffer data. If there are several shallow copies of shape, they all will have the same instances count, while it should be different from chunk to chunk. So large chunks and separate drawables.

psi29a commented:
> > A slight misnomer setting wise.
> 
> It is quite strange that _nearby_ groundcover starts to depend on _distant_ terrain.

The overarching strategy is to get OP/AG to the point that is always on since it's performance is better than having it off anyway. Then we can refactor many other parts of the terrain rendering away as we move forward.

Meaning that OP becomes the default and only terrain renderer, there will be no concept of distant terrain accept as an abstraction in people's mind with the only setting that matters being how far out do you render statics.

bosvensson1 commented:
> Actually, it is quite expected. Draw calls are quite expensive (at least, in OSG). That's why we have an Object Paging for active grid to merge objects to large shapes on the fly (instead of just drawing thousands of objects one by one).

 I was not suggesting the CPU bottleneck by itself to be surprising. What surprised me was the relative difference compared to not seeing those same bottlenecks for the terrain and objects. The groundcover is mostly clones of the same object everywhere while the terrain and objects use different textures. So any cpu bottlenecks we experience should be more likely to crop up with terrain and objects, in theory.

akortunov commented:
> Does anyone know why?

Initially such chunk size was selected to render just *terrain*, before Object Paging appeared. There was no need to tweak terrain chunks size since terrain rendering is relatively fast anyway.
From other hand, Object Paging tried to reuse chunks from different distances, what caused more pop-ins when larger chunks size was used.

bosvensson1 commented:
> The problem is that instances count is a part of shape buffer data (see osg::PrimitiveSet class). If there are several shallow copies of the same shape, they all will have the same instances count, while it should be different from chunk to chunk. So large chunks and separate drawables so far.

Correct, the drawable has to be unique for instancing. But most of its contents do not. Probably it is the unique State sets and state attributes that will have the largest impact as this is what osg uses to group drawables together during cull.

akortunov commented:
> while the terrain and objects use different textures

It is not actually true - there is a lot of cases when different meshes can use the same texture or stateset. Groundcover is not really different. If you use grass meshes with Object Paging with disabled shadows, you will get almost the same result (well, except of increased loading time and incleased RAM usage).

bosvensson1 commented:
> An only actual difference from common statics on mesh level is that most of grass meshes are transparent.

The transparency could be exacerbating the issue if set to use osg's transparent render bin. There the chunks are sorted into transparency rendering order which will obviously take more time for more chunks. I do not suppose the sorting is even necessary because we are not sorting the instances themselves. Most likely, they only need alpha testing not real transparency.

bosvensson1 commented:
> The transparency could be exacerbating the issue if set to use osg's transparent render bin.

It seems this is not the case in groundcover. There the render bin is manually set to inherit.

My point still holds with respect to OP AG as you presented OP AG's mediocre performance with groundcover as evidence that we should expect groundcover to be slow.

akortunov commented:
A short summary - **in its current shape** this PR brings some issues:
1. Pop-ins issue is back (because we can not just use stub data from *any* other distance instead of building an actual one without size effects). I'd consider it as a blocker and suggest to remove this change from this PR (because we do not know how to solve the issue properly so far). There can be a separate PR to fix this issue.
2. Added an implicit dependency between groundcover and distant terrain, what is not intuitive. Not a blocker, but definely not good.
3. Removed an ability to tweak chunks size, what may cause FPS drops on systems without high-end CPU. Probably it is worth to investigate a possibility to use different chunk size for everything instead of removing the setting, not just for groundcover.

akortunov commented on apps/openmw/mwrender/renderingmanager.cpp:
```diff
@@ -398,41 +400,27 @@ namespace MWRender
-            // Groundcover it is handled in the same way indifferently from if it is from active grid or from distant cell.
-            // Use a stub grid to avoid splitting between chunks for active grid and chunks for distant cells.
-            mGroundcoverWorld->setActiveGrid(osg::Vec4i(0, 0, 0, 0));
+            float groundcoverDistance = std::max(0.f, Settings::Manager::getFloat("rendering distance", "Groundcover"));
```
Initially distance multiplier was added to QuadTreeWorld to avoid pop-ins with large FOV, when it culled chunks which actually were visible. Did you test this case?

bosvensson1 replied:
The FOV is relevant for the far clipping plane. The points on the sides of the plane are further away than the view distance. Groundcover fading uses the true euclidean distance which ends in a sort of circle not a plane.

akortunov commented on components/terrain/quadtreeworld.cpp:
```diff
@@ -341,7 +341,7 @@ void loadRenderingNode(ViewData::Entry& entry, ViewData* vd, int vertexLodMod, f
         for (QuadTreeWorld::ChunkManager* m : chunkManagers)
         {
-            if (m->getViewDistance() && entry.mNode->distance(vd->getViewPoint()) > m->getViewDistance())
+            if (m->getViewDistance() && entry.mNode->distance(vd->getViewPoint()) > m->getViewDistance() + reuseDistance*10)
```
Probably it would be better to avoid magic numbers. It is unclear why 10 instead of, for example, 8 or 16.

bosvensson1 replied:
I agree but its not so straight forward. The real magic number to calibrate is the preloader loading time and prediction time. Will try to find a better solution in the future. It is not really specific to groundcover.

psi29a commented:
@bosvensson1 thanks to a recent change to quadtreeworld, making it sleepless, we have a merge conflict here. Can you have a look please?

bosvensson1 commented:
> A short summary - **in its current shape** this PR brings some issues:
> 
>     1. Pop-ins issue is back (because we can not just use stub data from _any_ other distance instead of building an actual one without size effects). I'd consider it as a blocker and suggest to remove this change from this PR (because we do not know how to solve the issue properly so far). There can be a separate PR to fix this issue.

The last commits attempt to plug this. If there are more cases this points to issues with terrain in general. Though we may not notice as much with the terrain as with groundcover.

>     2. Added an implicit dependency between groundcover and distant terrain, what is not intuitive. Not a blocker, but definely not good.

There is not much difference to master. Master creates the QuadTreeWorld that distant terrain uses too but neglects using it for the terrain. As I understand, there is no reason not to use it once we created it. It is definitely more efficient to use than the simple grid in my testing.

>     3. Removed an ability to tweak chunks size, what may cause FPS drops on systems without high-end CPU.

I did not remove the setting purposefully. There just is no straightforward way to tweak chunk size individually once we decide to share the QuadTreeWorld to reap the benefits of this PR.


> Probably it is worth to investigate a possibility to use different chunk size for everything instead of removing the setting, not just for groundcover.

Sure enough. This will touch more than groundcover though. I wish to keep scope nice and simple for my first PRs. I ran into enough weird crashes during the development that I am not comfortable to touch too many systems at once.

bosvensson1 commented:
> @bosvensson1 thanks to a recent change to quadtreeworld, making it sleepless, we have a merge conflict here. Can you have a look please?

This is new to me so I used the solution suggested by github. It looks alright to me besides the strange >>>> signs I had to remove. Sure hope this is the correct way.

psi29a commented:
> > @bosvensson1 thanks to a recent change to quadtreeworld, making it sleepless, we have a merge conflict here. Can you have a look please?
> 
> This is new to me so I used the solution suggested by github. It looks alright to me besides the strange >>>> signs I had to remove. Sure hope this is the correct way.

You're doing great! I've setup a sister MR so that we can get some coverage here and artefacts for people interested in testing:
https://gitlab.com/OpenMW/openmw/-/merge_requests/1205
We have an issue with windows not being able to copy the artefact over beause the `/` in the branch name; pfft...  will look into this later.

AbduSharif commented:
> There is not much difference to master. 
One big difference to master is the abrupt dependance on distant terrain being enabled for groundcover, which would neither make sense to the end user nor they're informed of that change in the documentation or settings-default.cfg

psi29a commented:
> > There is not much difference to master.
> 
> One big difference to master is the abrupt dependance on distant terrain being enabled for groundcover, which would neither make sense to the end user nor they're informed of that change in the documentation and settings-default.cfg, unless you still can have groundcover alone enabled and it will work and I misunderstood?

We should be getting things to the point that the end-user shouldn't be bothered with knowing the difference and having object paging on by default with only settings limiting its scope. I posted about this earlier here: https://github.com/OpenMW/openmw/pull/3096#issuecomment-914992830

The is simplify both in terms of code and the options to something that the end-user can reasonably understand. Right now, we have a lot of contradictory settings that don't make sense.

AbduSharif commented:
> We've already proven that having AG on improves performance, even on low-end machines, so why would anyone want it off?

Not really, the need for extra processing and power and more RAM could be avoided which would be a big help.

And anyway that means you still need to better reflect that in the documentation and settings-default.cfg for the end users.

psi29a commented:
> Not really, the need for extra processing and power and more RAM could be avoided which would be a big help especially when one doesn't really want to have a big viewing distance.

I was able to demonstrate this on a laptop from 2013 with a weak Intel iGPU and low-end mobile nvidia dGPU.

low-end machines are easily at or above 4GiB, if you want to start adding additional mods like groundcover and TR, then 8GiB is not unreasonable. 

We are NOT targeting 2001 PCs and if unless there is a dedicated Android developer on the OpenMW team, then Android and those devices will remain 2nd class citizens, otherwise it is hamstringing development.

> And anyway that means you still need to better reflect that in the documentation and settings-default.cfg for the end users that now distant terrain have to be enabled for groundcover, it's a new change than the usual.

Definitely, good thing we're still in the beginning of the 0.48 development cycle. :)

psi29a commented:
> Not really, the need for extra processing and power and more RAM could be avoided which would be a big help especially when one doesn't really want to have a big viewing distance.

I understand your point, but that is what we are trying to get to. Perhaps not in this PR, but in follow-ups.

> And anyway that means you still need to better reflect that in the documentation and settings-default.cfg for the end users that now distant terrain have to be enabled for groundcover, it's a new change than the usual.

Absolutely, as part of the process of removing cruft and simplifying terrain and object rendering, from the end-user perspective, they only thing they need worry about is how far out they want to view the horizon. From there OP can make decisions about how to optimally do that. 

I'm sure the ability to have, for example grass, only render in the active cells is more performant than seeing grass off into the horizon. It doesn't make much sense up to a point. We'll get there.

psi29a commented:
@bosvensson1 you still thinking on this or do you think it ready for re-review?

Can you update your OP to go over what this PR is doing now that it's gone through a few iterations or is still the same?

bosvensson1 commented:
There is no need to update the description. The few minor issues solved
since first opening the PR should not have existed to begin with. Yes,
this is ready for rereview. Everything pointed out to me has been answered
and addressed.

psi29a commented:
All in. Let's get this in for more testing.
Going to squash this to make it easy to revert if we need to.

AnyOldName3 commented:
I see lots of changes and discussion here since I last looked, yet it's been merged. That doesn't fill me with confidence.

> transparent

Grass is tested, not blended (well, techncially if you've got the setting enabled it's alpha-to-coverage) so it needs no depth sort as the Z-buffer handles it.

> forcing distant terrain on

Technically a more precise name for the setting would be `terrain paging` but normally we go for more human-friendly setting names, like *antialias alpha test* instead of *convert fixed function alpha test to shader discard or alpha to coverage depending on whether MSAA is on*. Under that approach, *object paging* should really have been called *distant statics* or *merged distant statics* or something, but that didn't happen. Anyway, people managed to misunderstnad *distant terrain* as *distant statics* for years, so it didn't even succeed at being human-friendly.

As long as we can't think of a good reason to have unpaged terrain with object paging or grass, I'm not really opposed to those things requiring it. Otherwise, maybe we should make the quad tree part implicit and make the setting only control terrain (therefore making it possible to have grass or object paging on with distant terrain off).

> There just is no straightforward way to tweak chunk size individually once we decide to share the QuadTreeWorld to reap the benefits of this PR.

😬

I'm fairly sure the optimal chunk size for simple instanced drawables is going to be bigger than for batches of statics. The bigger the batch, the greater the return from instancing (and the lower the per-element cost of the overhead), especially as the vertex shader is cheap and lots of instances are entirely clipped due to small triangles, whereas with big batches of statics, the drop in culling effectiveness will come back to bite you sooner.

> Going to squash this to make it easy to revert if we need to.

A merge commit also reverts cleanly in a single command that makes a single commit.

bosvensson1 commented:
>>    There just is no straightforward way to tweak chunk size individually once we decide to share the QuadTreeWorld to reap the benefits of this PR.

> 😬



I may have overstated. Diverging min chunk sizes are still possible with another approach. We can increase the min chunk size of the quad tree and add additional splits in the chunk managers. We may find this approach to be more efficient regardless of diverging min chunk sizes. Currently we often reload chunks because of miniscule changes in the organisation of the quad tree at the smallest scales. Leaving the last couple of splits up to the chunk managers should mitigate this effect. It should also reduce memory requirements because chunks are loaded on demand whereas the terrain world's quad tree is built for the whole world in advance and uses up exponential amounts of memory the smaller its min chunk size is.

I have not conducted these experiments so far because my next PR optimising the low level of groundcover is almost ready. With these optimisations the claimed difference between large and small chunks might disappear.

Sisah2 commented:
This commit actually decreased performance it seems, there is much more chunks rendered than before (even if using min chunk size = 0.125, same value as it use now), shaders hide it but with disabled fading and that thing that set vertex position to vec4(0,0,0,1), its clearly visible that much more grass is rendered. Second clue is number of groundcover chunks increased a lot in profiler (f4).

bosvensson1 commented:
This is to be expected, we accepted a temporary decrease in performance in exchange for other benefits. When compared to before this PR with min chunk size = 0.125 we can expect to see slightly more chunks than before because of the new reload threshold. With #3101 the impact of chunks will be reduced and you should see better performance again.

Sisah2 commented:
Those extra rendered chunks are far away of grass rendering distance, it cant be fixed? I did some screenshots on discord, not sure if you are here.

Sisah2 commented:
Ahh, seems like the difference in number of chunks decrease with distance, i was testing with 1000, after changing to 7000, the number of chunks is the same in my testing scene.

AbduSharif commented:
> This is to be expected, we accepted a temporary decrease in performance in exchange for other benefits.
What other benefits?

bosvensson1 commented:
> What other benefits?

Preloading, faster initial setup, code cleanup and a cleaner architecture that will benefit future improvements such as #3101 and the idea from comment https://github.com/OpenMW/openmw/pull/3096#issuecomment-916957950

> Ahh, seems like the difference in number of chunks decrease with distance, i was testing with 1000 (where were 10 chunks before this change and 29 after), after changing to 7000, the number of chunks is the same in my testing scene.

Correct. The reloading threshold is about 1000 units. Therefore it has a higher relative impact on the number of chunks created for lower rendering distances.

AnyOldName3 commented:
Just in case it clarifies some things, Sisah2 will be talking in the context of the Android port (sometimes specifically his fork of it) which uses GL4ES. Even when running on a device that supports instancing (e.g. it's got GLES 3, which 91.5% of still-in-use Android devices do according to https://developer.android.com/about/dashboards#OpenGL), GL4ES fakes instancing by making a flat vertex buffer with a copy of the shared data per instance. That's going to exacerbate any impact of recreating chunks frequently.

Sisah2 commented:
Yes, but those changes are not related to paging/instancing the base of those two seems to be the same. 
But wanna ask what enabled preloading exactly mean?
Does it mean that chunks are now created on preload thread and before were made on draw thread? Or just that meshes are now properly preloaded and before it loaded them only when new chunk was created, so caused small stuter?

bosvensson1 commented:
> But wanna ask what enabled preloading exactly mean?

It means the groundcover chunks are now part of preloading features described in settings.cfg. Creating a groundcover chunk involves scanning through content files to find groundcover instances and create them according to criteria such as player position and view distance. This was previously done in the main thread and should now done in the background. In practice you should see fewer frame drops.

glassmancody commented:
Wireframe mode no longer works, issue is tracked here https://gitlab.com/OpenMW/openmw/-/issues/6290
